File: ecmascript/beer-song/beer-song.js
class Beer {

  static verse(num){
    let v;

    if(num === 0){
      v = `No more bottles of beer on the wall, no more bottles of beer.
Go to the store and buy some more, 99 bottles of beer on the wall.
`
    }else if(num === 1){
      v =`1 bottle of beer on the wall, 1 bottle of beer.
Take it down and pass it around, no more bottles of beer on the wall.
`;
    }else if(num === 2){
      v = `${num} bottles of beer on the wall, ${num} bottles of beer.\nTake one down and pass it around, ${num -1} bottle of beer on the wall.\n`
    }else{
      v = `${num} bottles of beer on the wall, ${num} bottles of beer.\nTake one down and pass it around, ${num -1} bottles of beer on the wall.\n`
    }
    return v;
  }

  static sing(start = 99, end = 0){
    let s = "";

    for(; start >= end; start--){
      s = s + Beer.verse(start);
      if(start !== end){
        s = s + "\n";
      }
    }

    return s;
  }
}

export default Beer

File: elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(number) do

    if number == 0 do
      "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
    else

      plural = fn
        n when n > 1 -> "s"
        n when n == 1 -> ""
        n when n <= 0 -> "s"
      end

      one_or_it = if number > 1, do: "one", else: "it"
      i_or_nomore = if number > 1, do: "#{number - 1}", else: "no more"

      "#{number} bottle#{plural.(number)} of beer on the wall, #{number} bottle#{plural.(number)} of beer.\nTake #{one_or_it} down and pass it around, #{i_or_nomore} bottle#{plural.(number - 1)} of beer on the wall.\n"
    end
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range) do
    s = Enum.reduce(range, "", fn i, acc -> acc <> verse(i) <> "\n" end)
    String.slice(s, 0..-2)
  end

  def lyrics() do
    lyrics(99..0)
  end
end
